DigitalLearningSolutions.Web/Controllers/TrackingSystem/CourseSetup/CourseSetupController.cs

```diff
@@ -191,27 +200,36 @@ public IActionResult SelectCourseAllCourses()
                     .Single(ap => ap.ApplicationId == applicationId);
 
             data!.SetApplicationAndResetModels(selectedApplication);
-            TempData.Set(data);
+
+            multiPageFormService.SetMultiPageFormData(data, MultiPageFormDataFeature.AddNewCourse, TempData);
 
             return RedirectToAction("SetCourseDetails");
         }
 
-        [ServiceFilter(typeof(RedirectEmptySessionData<AddNewCentreCourseData>))]
         [HttpGet("AddCourse/SetCourseDetails")]
+        [ServiceFilter(typeof(RedirectMissingMultiPageFormData<AddNewCentreCourseData>))]
         public IActionResult SetCourseDetails()
         {
-            var data = TempData.Peek<AddNewCentreCourseData>();
+            var data = multiPageFormService.GetMultiPageFormData<AddNewCentreCourseData>(
+                MultiPageFormDataFeature.AddNewCourse,
+                TempData
+            );
 
-            var model = data?.SetCourseDetailsModel ?? new SetCourseDetailsViewModel(data!.Application!);
+            var model = data.CourseDetailsData != null
+                ? new SetCourseDetailsViewModel(data.CourseDetailsData)
+                : new SetCourseDetailsViewModel(data!.Application!);

```

Comment:
By following the normal process of the journey? It's set on the first page of the journey. If we skip to here without actually setting it, I think throwing here because it is null is acceptable.

Reply:
It'd be nice to do that explicitly I think. If I came across this code in a year, I'd just assume that either the dev who wrote it noticed something I didn't or just didn't think about what would happen if it were null. Whereas if there's an explicit "if this is null then throw" block, I know that the dev has thought about it and deliberately chosen the appropriate outcome.